### PR TITLE
and, or, &&, ||

### LINKED ISSUE #137: Equivalent to bash's "&&"
Great project! It would be very nice to have a keyword in xonsh that is equivalent to `&&` in bash/zsh or `;and` in fish, meaning that it would execute \<B\> if and only if \<A\> is successful in the following example line:
`<A> <keyword> <B>`.
For maximum "bashwards compatibility" the keyword should be exactly like in bash (&&) if possible.

For completeness' sake there should also be a keyword equivalent to `||` (bash/zsh) or `;or` (fish) that only executes something if the preceding command fails.

The current xonsh behavior (v0.1.2) for this syntax seems to be 100% cpu usage and doing nothing until I hit Ctrl+C

### PR DESCRIPTION
**Pull request by scopatz**
The much requested, long delayed subprocess logical operators have arrived. This should close #137 and #434, finally. This builds on #671.

Though I much dreaded implementing this, it really wasn't so bad once all the pieces were in place.

I'd really appreciate people banging on this to get corner cases I might not have thought of. 

The next release after this is merged should be v0.3.0 as it does modify the language somewhat.

### CODE REVIEW

**Comment by adqm** [comment]
(Back from the dead)
I don't have time to test in depth yet, but I had a couple of thoughts:

If you remember, I had a couple of half-finished attempts at this in the past.  In both of those cases, I had this basic functionality working (plus grouping via parentheses), but where things broke down was dealing with IO redirection and job control.  After a couple of quick tests, it looks like those features aren't implement as of yet.

One question in my mind is whether it makes sense to try to get those two things working first, or whether we should put this in as is (with the understanding that getting those things working may involve some large amount of rewriting and restructuring).

**Comment by scopatz** [comment]
Hi @adqm!

> If you remember, I had a couple of half-finished attempts at this in the past. In both of those cases, I had this basic functionality working (plus grouping via parentheses), but where things broke down was dealing with IO redirection and job control. After a couple of quick tests, it looks like those features aren't implement as of yet.

yep, I remember.  I hope I am not stepping on your toes by putting this in.  The features you mention *have not* been implemented yet, as you say.

> One question in my mind is whether it makes sense to try to get those two things working first, or whether we should put this in as is (with the understanding that getting those things working may involve some large amount of rewriting and restructuring).

So my idea and preference here is to put this in as is and break up the other work into their own PRs.  I think going for the monlithic PR coupled with the prior, overly complex parser was what has prevented this feature from really moving forward.

I also think that 90% (to pick a big number) of what people actually want to do with && and || is very simple. They just want to `and` two subprocs together and they have been disappointed when this wasn't available. This PR is really just trying to address that need.

For the remaining 10% (to pick a smaller number), I think it is better to say, "we know, it isn't fully implemented, but the basics are there...please help if you can."  This is very different than what we say right now, which is, "yeah, people have been asking for it and it is on our radar. We'll get to it eventually."

I am trying to approach my xonsh efforts right now mostly from a what do we need for (what is blocking) 1.0 point of view.  I am not saying that this is complete or that 1.0 will happen because of this, but I think it was one of the big missing pieces. We probably should have a v1.0 tracking issue to keep tabs on progress.  We can always delay 1.0 for QA reasons, but I'd really like to get all of the top-level pieces in place. Just my two drachmas.

**Comment by adqm** [comment]
>  I hope I am not stepping on your toes by putting this in

Certainly not!

As for your other points, that works fine for me!  I agree on most points.  I won't have a chance to test right now, but I should be able to give this a thorough look soon.  I may also try to hack in grouping by parentheses when I do take a look, as that seems to me to be part of the "core" feature (with job control and IO redirect farther down the list).

**Comment by scopatz** [comment]
Sounds great to me!

**Comment by scopatz** [comment]
FIxed merge conflict with master such that this now includes the color stuff.

**Comment by adqm** [comment]
I am getting exceptions from `run_subproc` with this:
```
hartz@lappy-514 ~/xonsh $ ls and ls
appveyor.yml  build          CONTRIBUTING  license   MANIFEST.in        README.rst  release.xsh       scripts   tests       xonsh
binstar.yml   CHANGELOG.rst  docs          logo.txt  ply-3.8-py3.4.egg  recipe      requirements.txt  setup.py  travis.yml  xonsh.egg-info
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/site-packages/xonsh/base_shell.py", line 150, in default
    self.execer.exec(code, mode='single', glbs=self.ctx)  # no locals
appveyor.yml  build          CONTRIBUTING  license   MANIFEST.in        README.rst  release.xsh       scripts   tests       xonsh
binstar.yml   CHANGELOG.rst  docs          logo.txt  ply-3.8-py3.4.egg  recipe      requirements.txt  setup.py  travis.yml  xonsh.egg-info
  File "/usr/local/lib/python3.5/site-packages/xonsh/execer.py", line 126, in exec
    return exec(code, glbs, locs)
  File "<xonsh-code>", line 1, in <module>
  File "/usr/local/lib/python3.5/site-packages/xonsh/built_ins.py", line 702, in subproc_uncaptured
    return run_subproc(cmds, captured=False)
  File "/usr/local/lib/python3.5/site-packages/xonsh/built_ins.py", line 685, in run_subproc
    if _wait_for_proc(prev_proc, prev_is_proxy, procs, cmds, background):
  File "/usr/local/lib/python3.5/site-packages/xonsh/built_ins.py", line 476, in _wait_for_proc
    proc.stdout.close()
AttributeError: 'NoneType' object has no attribute 'close'
```

**Comment by scopatz** [comment]
Sorry @adqm, this should be fixed now.

**Comment by scopatz** [comment]
Any objections to this going in?

File: docs/tutorial.rst [hunk 1]
```diff
@@ -416,6 +416,59 @@ Python operator.
 If you are unsure of what pipes are, there are many great references out there.
 You should be able to find information on StackOverflow or Google.
 
+Logical Subprocess And
+=======================
+
+Subprocess-mode also allows you to use the ``and`` operator to chain together 
+subprocess commands. The truth value of a command is evaluates as whether
```
Comment (gforsyth) [inline docs/tutorial.rst]: *evaluated

File: docs/tutorial.rst [hunk 2]
```diff
@@ -416,6 +416,59 @@ Python operator.
...
+its return code is zero (i.e. ``proc.returncode == 0``).  Like in Python, 
+if the a command evaluates to ``False``, subsequent commands will not be executed.
```
Comment (gforsyth) [inline docs/tutorial.rst]: 'a command' -> 'command'

**Comment by gforsyth** [comment]
Barring those two doc-tweaks, this works as expected for me.  I was having the same error as @adqm before your fixes.

**Comment by scopatz** [comment]
Thanks for spotting those @gforsyth. I have pushed up some fixes.

**Comment by moigagoo** [comment]
Hi!

Apparently, this broke running scripts in background with a trailing &.

Used to be:
```
$ sci start &
[1] 20026
$ Starting Sloth CI on http://localhost:8080
$ #not blocking
```

Now:
```
$ sci start &
$ Starting Sloth CI on http://localhost:8080 # blocking
```

**Comment by scopatz** [comment]
Hi @moigagoo, I am a little confused. This hasn't been merged yet so do you mean this branch is broken or master is broken?

**Comment by scopatz** [comment]
Ahh nm.  I just tried it with `sleep 10 &`, and it is this branch that is broken.

**Comment by moigagoo** [comment]
@scopatz I'm testing andor now, that's how I faced the issue. Sorry for the confusion.

**Comment by zommerfelds** [comment]
Is this being worked on? && is a widely used feature in Bash, I would love to have this in xonsh.

**Comment by scopatz** [comment]
Yes, this is, though help would be most welcome. Note that right now you can do `![...] and ![...]` instead.

**Comment by scopatz** [comment]
Now see #789

**Pull request merged**